misc/validate_accounting.py

```diff
@@ -599,139 +548,212 @@ def token_addr_from_args(args):
+    if algod_account['round'] != indexer_account['round']:
+        indexer_account = indexerAccountFromAddr(indexer, niceaddr, blockround=algod_account['round'], headers=indexer_headers)
+    i2a_checker.check(indexer_account, algod_account)
+
+def compare_thread(ia_queue, i2a_checker, algod, indexer, indexer_headers):
+    while True:
+        indexer_account = ia_queue.get()
+        if indexer_account is None:
+            return
+        compare(indexer_account, i2a_checker, algod, indexer, indexer_headers)
+
+# decrement by one the big-endian number implicit in a byte string
+# crashes on [0,0,0, ...]
+def bytedec(bs):
+    bs = list(bs)
+    pos = len(bs) - 1
+    while bs[pos] == 0:
+        bs[pos] = 0xff
+        pos -= 1 # carry from previous position
+    bs[pos] -= 1
+    return bytes(bs)
+
+# for --shard a/b return address range (minaddr-1, maxaddr)
+# a in [1..b], e.g. --shard 1/16 .. 16/16
+# may return (None, addr) or (addr, None) for first/last shard
+def shard_bounds(shardspec):
+    a,b = shardspec.split('/')
+    a = int(a)
+    b = int(b)
+    chunk = (2**64) // b
+    if a < 1:
+        a = 1
+    if a > b:
+        a = b
+
+    if a == 1:
+        minsplit = None
+    else:
+        minsplit = struct.pack('>Q', chunk * (a-1))
+        minsplit += bytes([0]*24)
+        minsplit = bytedec(minsplit)
+    if a == b:
+        maxsplit = None
+    else:
+        maxsplit = struct.pack('>Q', chunk * a)
+        maxsplit += bytes([0]*24)
+    return (minsplit, maxsplit)
+
 def check_from_algod(args):
-    gpath = args.genesis or os.path.join(args.algod, 'genesis.json')
-    getGenesisVars(gpath)
+    global graceful_stop
     token, addr = token_addr_from_args(args)
     algod = AlgodClient(token, addr)
+    if args.genesis or args.algod:
+        gpath = args.genesis or os.path.join(args.algod, 'genesis.json')
+        getGenesisVars(gpath)
+    else:
+        process_genesis(algod.algod_request("GET", "/genesis"))
     status = algod.status()
-    #logger.debug('status %r', status)
     iloadstart = time.time()
-    if args.indexer:
-        i2a = indexerAccounts(args.indexer, addrlist=(args.accounts and args.accounts.split(',')))
-        i2a_checker = CheckContext(i2a, sys.stderr)
-    else:
-        logger.warn('no indexer, nothing to do')
-        return None, None
-    iloadtime = time.time() - iloadstart
-    logger.info('loaded %d accounts from indexer in %.1f seconds, %.1f a/s', len(i2a), iloadtime, len(i2a)/iloadtime)
-    aloadstart = time.time()
-    rawad = []
-    for address in i2a.keys():
-        niceaddr = algosdk.encoding.encode_address(address)
-        rawad.append(algod.account_info(niceaddr))
-    aloadtime = time.time() - aloadstart
-    logger.info('loaded %d accounts from algod in %.1f seconds, %.1f a/s', len(rawad), aloadtime, len(rawad)/aloadtime)
+    indexer_headers = None
+    gtaddr = args.gtaddr
+    maxaddr = None
+    if args.shard:
+        minaddr, maxaddr = shard_bounds(args.shard)
```

Comment:
I found that postgres could dump all 8+ million accounts in 8 seconds, so this stuff may not be necessary. Just grab all the accounts and divvy them up.